File: Sheets/src/neetcode/practice/Two_Pointers/_5_trapRainWater.java
package neetcode.practice.Two_Pointers;

public class _5_trapRainWater {
  public static int trap(int[] height) {
    if (height == null || height.length == 0) return 0;

    int n = height.length;
    int l = 0, r = n - 1;

    int leftMaxHeight = height[l], rightMaxHeight = height[r];
    int ans = 0;

    while (l < r) {
      if (leftMaxHeight < rightMaxHeight) {
        l++;
        leftMaxHeight = Math.max(leftMaxHeight, height[l]);
        ans += leftMaxHeight - height[l]; // Would be no change if this was a new maximum
      } else {
        r--;
        rightMaxHeight = Math.max(rightMaxHeight, height[r]);
        ans += rightMaxHeight - height[r]; // Would be no change if this was a new maximum
      }
    }

    return ans;
  }
}

File: Sheets/src/neetcode/practice/Two_Pointers/_5_TrapRainWater.cc
#include <vector>

class Solution
{
public:
  int trap(std::vector<int> &height)
  {
    if (height.empty())
      return 0;

    std::size_t n = height.size();
    std::size_t l{0}, r{n - 1};

    int left_max_height{height[l]}, right_max_height{height[r]};
    int ans{};
    while (l < r)
    {
      if (left_max_height < right_max_height)
      {
        l++;
        left_max_height = std::max(left_max_height, height[l]);
        ans += left_max_height - height[l];
      }
      else
      {
        r--;
        right_max_height = std::max(right_max_height, height[r]);
        ans += right_max_height - height[r];
      }
    }

    return ans;
  }
};
